src/primitives.jl

```diff
@@ -80,36 +80,35 @@ function forward!(i::BroadcastInstruction)
     output_variable = isa(i.output, Variable) ? i.output : first(i.output)
     f, input_values = first(i.input), value.(i.input[2:end])
     output_value = value(output_variable)
-    input_derivs = similar(output_value, size(output_value, 1), size(output_value, 2), length(input_values))
+    input_derivs = map(_ -> similar(output_value), input_values)
     dual_eval_broadcast!(output_value, input_derivs, f, input_values)
     i.output = (output_variable, input_derivs)
     return nothing
 end
 
-# Use ForwardDiff to calculate `f(values[1][i], values[2][i], ...)` and
-# `∇f(values[1][i], values[2][i], ...)` from a single elementwise application. This
-# implementation is actually incomplete, but it doesn't matter for our performance
-# experiment. Specifically, it doesn't implement the proper reduction/expansion semantics
-# encountered when the arguments have different shapes. In other words, this implementation
-# only works when all broadcast arguments are arrays of the same shape (which is sufficient
-# for our benchmarking purposes).
 @noinline function dual_eval_broadcast!(output_value::AbstractMatrix,
-                                        input_derivs::AbstractArray{<:Any,3},
+                                        input_derivs::NTuple{N,<:AbstractMatrix},
                                         kernel,
                                         input_values::NTuple{N,<:AbstractMatrix}) where {N}
     @assert all(size(iv) === size(output_value) for iv in input_values)
-    for i in 1:size(output_value, 1)
-        for j in 1:size(output_value, 2)
-            ij_result = dual_eval_kernel(kernel, getindex.(input_values, i, j))
-            output_value[i, j] = ForwardDiff.value(ij_result)
-            for k in 1:N
-                input_derivs[i, j, k] = ForwardDiff.partials(ij_result, k)
-            end
-        end
+
+    # Use ForwardDiff's `Dual` numbers to calculate `kernel.(input_values...)` and
+    # elementwise derivatives of `kernel` at the same time (`output_duals` is an array
+    # of dual numbers).
+    output_duals = @fastsplat(broadcast((args...) -> dual_eval(kernel, args), input_values...))
```

Comment:
So the key tradeoff here is that this approach requires an extra allocation for `output_duals` over the previous approach.

However, despite the name of the PR, I'm wondering if we still need to preserve the SoA approach in light of the recent improvements we've seen due to more aggressive inlining. If not, then we could just store `output_duals` here and get rid of the (theoretically redundant) `input_derivs`. Then, we could go back to doing the load in the backwards pass (I'll leave another comment below with details).